### ISSUE TITLE
ArcFace loss code: the meaning of mm

### ISSUE THREAD

**Issue by ghost** [deepinsight/insightface#247]
In the ArcFace code 

> mm = math.sin(math.pi-m)*m

 line present.
Could you explain what is the meaning/purpose of using mm? It seems that it lowers logit of ground truth class when angle between weight and sample + margin is bigger than pi (cosine function out of monotonic range and we can't apply margin directly), but why it takes this particular form?
Also why is it valid to multiply something in sine space (and actually why sin?) and margin which is in angular space?
Thanks for helping!

**Comment by samxuxiang** [comment]
I am curious to know the performance when the model is trained strictly following the loss function presented in the original paper.

**Comment by HoiM** [comment]
Yes. Maybe it is kind of hard for the model to converge if the original loss function is deployed. So some tricks are needed to ease the learning process.

**Comment by Jaimin7632** [comment]
Still, the rationale behind how this particular trick makes it easier for the model to converge is unknown. Please make it clear what is the mathematics behind using the below statements:

`mm = math.sin(math.pi - m) * m`
`zy_keep = zy - s * mm`

- What is the purpose behind taking zy_keep as `zy - s * m * sin(m)` ?
